File: src/2019/day1.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2019/day/1
    std.debug.print("Day 1, 2019: The Tyranny of the Rocket Equation\n", .{});

    const m = utils.Monitor.init();
    defer m.deinit();
    const input = comptime utils.trim_input(@embedFile("input"));

    const p1 = try part1(input);
    std.debug.print("  Part 1: {}\n", .{p1});
    std.debug.assert(3538016 == p1);

    const p2 = try part2(input);
    std.debug.print("  Part 2: {}\n", .{p2});
    std.debug.assert(5304147 == p2);
}

fn getFuel(mass: i32) i32 {
    return @divFloor(mass, 3) - 2;
}

fn part1(input: []const u8) !i32 {
    var sum: i32 = 0;
    var it = std.mem.tokenizeAny(u8, input, "\r\n");
    while (it.next()) |line| {
        sum += getFuel(try std.fmt.parseInt(i32, line, 10));
    }

    return sum;
}

test "part 1" {
    comptime try std.testing.expectEqual(2, getFuel(12));
    comptime try std.testing.expectEqual(2, getFuel(14));
    comptime try std.testing.expectEqual(654, getFuel(1969));
    comptime try std.testing.expectEqual(33583, getFuel(100756));
}

fn getRealFuel(mass: i32) i32 {
    var fuel = getFuel(mass);

    if (fuel > 0) {
        return fuel + getRealFuel(fuel);
    }

    if (fuel < 0) {
        fuel = 0;
    }

    return fuel;
}

fn part2(input: []const u8) !i32 {
    var sum: i32 = 0;
    var it = std.mem.tokenizeAny(u8, input, "\r\n");
    while (it.next()) |line| {
        sum += getRealFuel(try std.fmt.parseInt(i32, line, 10));
    }

    return sum;
}

test "part 2" {
    comptime try std.testing.expectEqual(2, getRealFuel(12));
    comptime try std.testing.expectEqual(2, getRealFuel(14));
    comptime try std.testing.expectEqual(966, getRealFuel(1969));
    comptime try std.testing.expectEqual(50346, getRealFuel(100756));
}

File: src/2019/day1.cpp
#include "day1.hpp"

#include "../common.hpp"

#include <fstream>

constexpr int GetFuel(int mass)
{
    return mass / 3 - 2;
}

static_assert(2 == GetFuel(14));
static_assert(2 == GetFuel(14));
static_assert(654 == GetFuel(1969));
static_assert(33'583 == GetFuel(100'756));

constexpr int GetRealFuel(int mass)
{
    int fuel = GetFuel(mass);

    if (fuel > 0)
    {
        return fuel + GetRealFuel(fuel);
    }

    if (fuel < 0)
    {
        fuel = 0;
    }

    return fuel;
}

static_assert(2 == GetRealFuel(12));
static_assert(2 == GetRealFuel(14));
static_assert(966 == GetRealFuel(1969));
static_assert(50'346 == GetRealFuel(100'756));

int main()
{
    std::println("Day 1, 2019: The Tyranny of the Rocket Equation");
    auto const input = ParseInputNumbers();

    int const part1 = std::accumulate(begin(input), end(input), 0,
        [](int r, int mass)
        {
            return r + GetFuel(mass);
        });

    int const part2 = std::accumulate(begin(input), end(input), 0,
        [](int r, int mass)
        {
            return r + GetRealFuel(mass);
        });

    std::println("  Part1: {}", part1);
    Assert(3'538'016 == part1);

    std::println("  Part2: {}", part2);
    Assert(5'304'147 == part2);
}
